Skip/fix tests on OTP 19

antipax opened the pull request:
Addresses #4851.

In addition to the original issues from #4851, a test for `IEx` related to the `h` helper and callback functions was failing. I am currently skipping that test, but we can update `h` to use a similar approach to `callbacks_for_beam` in `elixir/test_helper.exs` as a fallback as part of this PR. That way it will continue to work for in memory and disk modules in <19, but it will only work for disk modules in >19.

I've also posted to erlang-questions asking why callbacks are no longer in the attributes, since I cannot find any reference to such a change in the changelog for 19, the github, or the erlang issue tracker. It seems this may be an unexpected change?

/cc @josevalim

fishcakez commented:
This doesn't look like "make tests pass" but rather "don't run tests that fail". Could we split this into a fix for callbacks and a commit that skips dialyzer tests?

antipax commented:
Sure.
On Jun 22, 2016 5:53 PM, "James Fish" <notifications@github.com> wrote:

> This doesn't look like "make tests pass" but rather "don't run tests that
> fail". Could we split this into a fix for callbacks and a commit that skips
> dialyzer tests?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/elixir-lang/elixir/pull/4855#issuecomment-227888734>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AADcogDZqwqZTnXz3Vy2yQSUOx7BDxt4ks5qOa71gaJpZM4I8OnO>
> .
>

josevalim commented on lib/iex/test/iex/helpers_test.exs:
```diff
@@ -71,6 +71,7 @@ defmodule IEx.HelpersTest do
     cleanup_modules([Sample])
   end
 
+  @tag :does_not_apply_to_otp19
```
I think we need to make this test pass properly.

antipax replied:
The only issue is that on OTP 19 the module will be in memory and thus we cannot get the abstract code in the h helper.

antipax replied:
Oh, wait. I see that this test actually compiles the files to beams in this test. Should be possible to make it work then.

josevalim commented on lib/elixir/test/elixir/test_helper.exs:
```diff
@@ -1,4 +1,10 @@
-ExUnit.start [trace: "--trace" in System.argv]
+exclude =
+  case :erlang.system_info(:otp_release) do
```
I think we can do this:

```elixir
otp_release = List.to_integer :erlang.system_info(:otp_release)
ExUnit.start [exclude: [:otp_release], include: [otp_release: otp_release], trace: "--trace" in System.argv]
```

And for the tests that do not work on 19, we can do: `@tag otp_release: 18`.

What do you think?

fishcakez replied:
Will we continue to skip the tests if this get patched before OTP 20?

antipax replied:
Seems the functional difference would be that then tests will be disabled on versions before and after 18, vs. the current approach which only disables the tests on 19. I think we'd rather keep the tests running on previous versions (for anyone still developing on 17, unless that is no longer supported?) and ensure that the disabled tests run again on any new versions?

antipax replied:
We would have to manually remove the skip for 19.1 for instance, in any event.

fishcakez replied:
We wont be able to compile on 17 anymore so dont worry about that.

fishcakez replied:
To be clear the `v1.1` branch is the last time we supported 17.

antipax replied:
Alternately, we could use `:erlang.system_info(:version)` which returns the erts version, which does include information about point releases.

fishcakez replied:
> We would have to manually remove the skip for 19.1 for instance, in any event.

How do we plan to do that?

antipax replied:
Ah... I had thought at least 1.2 supported 17.

fishcakez replied:
Would it be better to use the version of the dialyzer application instead?

antipax replied:
@fishcakez when the fix is released we would remove the exclusion? Or again, we could use system_info(:version), but I'm not sure that's guaranteed to increment with new minor releases.

antipax replied:
@fishcakez Yeah, that probably makes the most sense.

josevalim replied:
Yes, one step at a time. And this are tests, we can afford to be flexible and change later on.

fishcakez replied:
For future cases like this I think we should stick to using application versions because people can build custom versions of OTP that don't match an OTP version.

fishcakez replied:
> Yes, one step at a time. And this are tests, we can afford to be flexible and change later on.

OK lets stick with `:otp_release` I was just trying to drive home the point that this kind of version checking is often inadequate.

antipax replied:
Ok, so I'm gonna stick with it as is so that the tests start running again on OTP 20? Does that sound ok?

josevalim replied:
I really don't like the current tag name though. Can we use a Boolean flag
for :otp19 then? And we exclude based on true and false.

On Thursday, June 23, 2016, Eric Entin <notifications@github.com> wrote:

> In lib/elixir/test/elixir/test_helper.exs
> <https://github.com/elixir-lang/elixir/pull/4855#discussion_r68151089>:
>
> > @@ -1,4 +1,10 @@
> > -ExUnit.start [trace: "--trace" in System.argv]
> > +exclude =
> > +  case :erlang.system_info(:otp_release) do
>
> Ok, so I'm gonna stick with it as is so that the tests start running again
> on OTP 20? Does that sound ok?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elixir-lang/elixir/pull/4855/files/1bd0b165454dd143d26c34444bdb599faa942b2a#r68151089>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAAlbjt2iwzcU19WfKbQVrDCCB6H1CI4ks5qOb-3gaJpZM4I8OnO>
> .
>


-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

antipax replied:
That's fine. I'm not married to the name. :)

antipax commented:
Okey dokey, I think we're good to go now. I switched to using `Kernel.Typespec.beam_callbacks/1` since it does the same thing as my helpers did. I also switched to that function in code underlying the `h` helper, which was already using the function in other parts of its implementation, so this should not actually change what modules it was able to introspect. 

Additionally, I identified a potential bug (looks like `h` for a m/f/a that implemented a callback could have displayed the callback docs of the wrong arity, if there was more than 1 callback with the given name), and added a test for that case.

josevalim commented:
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:
